### PR TITLE
Nano: allow pipeline to specify CPU cores for each stage

### PR DESCRIPTION
**Pull request by MeouSker77**
## Description

Improve pipeline: Allow pipeline to specify CPU cores for each stage.
And fix nano nightly test: windows doesn't support pickle local object

### 1. Why the change?

For better performance

### 2. User API changes

```python
from bigdl.nano.pytorch import Pipeline

pipeline = Pipeline([
    ("preprocess", preprocess, {"core_num": 4}),
    ("inference", inference, {"core_num": 8})
])
```

### 3. Summary of the change 

- Allow pipeline to specify CPU cores for each stage.
- Add a new utils `EnvContext` to simplify setting environment variables temporarily.

### 4. How to test?

- [ ] Unit test

### CODE REVIEW

File: python/nano/src/bigdl/nano/utils/common/env.py [hunk 1]
```diff
@@ -62,3 +62,37 @@ def _find_library(library_name: str, priority_dir: Union[str, None] = None) -> U
...
+        if self.new_env is not None:
+            os.environ.update(self.new_env)
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        # it seems `os.environ = self.env_backup` will cause bug
+        for k, v in self.env_backup.items():
+            if k not in os.environ:
+                os.environ.pop(k)
+            else:
+                os.environ[k] = v
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/utils/common/env.py:98]: does this exit pop out those key only in `new_env`?

Comment (MeouSker77) [reply python/nano/src/bigdl/nano/utils/common/env.py:98]: oh, it's a bug, I'll fix it

Comment (MeouSker77) [reply python/nano/src/bigdl/nano/utils/common/env.py]: fixed it

**Review by rnwang04 — APPROVED** [review]
LGTM

File: python/nano/src/bigdl/nano/pytorch/inference/pipeline.py [hunk 2]
```diff
@@ -91,10 +97,23 @@ def run(self, inputs):
...
+        if isinstance(config.get("core_num", None), int):
+            core_num = config["core_num"]
+            cores = self.cores[:core_num]
+            self.cores = self.cores[core_num:]
+            if len(cores) < core_num:
+                warnings.warn(f"stage {name} requires {core_num} cores,"
+                              f" but there are only {len(cores)} cores left")
+            subprocess_env["KMP_AFFINITY"] = (f"granularity=fine,proclist"
+                                              f"=[{','.join([str(i) for i in cores])}],explicit")
+            subprocess_env["OMP_NUM_THREADS"] = str(len(cores))
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/pytorch/inference/pipeline.py:110]: I think we may just allow users to state the environment variables themselves. `core_num` seems to be a good shortcut for users but shortcut always has some limitations (e.g., this `core_num` only controls the openmp workload's thread num with a specific affinity strategy)

File: python/nano/src/bigdl/nano/pytorch/inference/pipeline.py [hunk 3]
```diff
@@ -42,13 +47,13 @@ def __init__(self, stages: List[Tuple]):
...
-                ("inference", model, {}),
+                ("preprocess", preprocess, {"core_num": 4}),
+                ("inference", model, {"core_num": 8}),
             ])
         ```
         will create a pipeline which has stage "preprocess" and "inference",
-        and the "preprocess" stage will call `preprocess(input)`, while the
-        "inference" stage will call `model(input)`.
+        and the "preprocess" stage will call `preprocess(input)` with 4 CPU cores,
+        while the "inference" stage will call `model(input)` with 8 CPU cores.
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/pytorch/inference/pipeline.py:56]: +"for openmp workload" maybe

**Review by TheaperDeng — APPROVED** [review]

File: python/nano/src/bigdl/nano/pytorch/inference/pipeline.py (hunk 2 above)
Comment (MeouSker77) [reply python/nano/src/bigdl/nano/pytorch/inference/pipeline.py:110]: We may provide another key named `env` for users to override the environment variables of subprocess ? e.g.
`pipeline = Pipeline([("preprocess", preprocess, {"core_num": 4, "env": {"KMP_AFFINITY": "granularity=fine,none"}}])`

File: python/nano/src/bigdl/nano/pytorch/inference/pipeline.py (hunk 3 above)
Comment (MeouSker77) [reply python/nano/src/bigdl/nano/pytorch/inference/pipeline.py]: done

**Comment by yangw1234** [comment]
Do we have documentation on when and how to use the `pipeline` feature?

**Comment by MeouSker77** [comment]
no, I think it's still in a very early state

> Do we have documentation on when and how to use the `pipeline` feature?

**Pull request merged**